Added multithreading support for mkcomposefs

divineaugustine opened the pull request:
This resolves #249 and results in 10x improvement for digest calculation and file copy.

Please refer the below document for details about the issue and the change. I would like to know your thoughts.
[threading.md](https://github.com/containers/composefs/files/14743119/threading.md)

cgwalters commented on libcomposefs/lcfs-writer.c:
```diff
@@ -471,9 +471,16 @@ struct lcfs_node_s *lcfs_node_new(void)
+	node->content = NULL;
 	node->ref_count = 1;
 	node->inode.st_nlink = 1;
+	node->delayedDigestCalculation = false;
```
Naming scheme in this project for member variables is `snake_case`, not `camelCase`.

divineaugustine replied:
I have corrected it here and in rest of my changes.

cgwalters reviewed (commented):
Thank you so much for working on this!  This may take some time to review as multithreading and C is tricky.

Can you please run `clang-format`?  Basically let's get past the superficial bits before we can dig into the threading.

cgwalters reviewed (dismissed):
Marking requested-changes to start just for formatting/style

divineaugustine commented:
> Can you please run `clang-format`? Basically let's get past the superficial bits before we can dig into the threading.
Thank you for checking. I have run clang-format on my changes via vscode formatting. Is that what you meant? 
FYI: I did rebase and force push to fix DCO error.

divineaugustine commented:
Sorry for the inconvenience one header file was missed out from formatting. I have corrected that as well. Hopefully this solves all the formatting issues from my changeset.

alexlarsson commented:
I'll have a look at the code, but can you please rebase this and squash the fixups so the end result is a minimal set of independent changes. Say one for the liibrary changes, and one of the mkcomposefs use of it.

alexlarsson commented on libcomposefs/lcfs-writer.c:
```diff
@@ -471,9 +471,16 @@ struct lcfs_node_s *lcfs_node_new(void)
 	}
-
+	
+	memset( node, 0, sizeof(struct lcfs_node_s));
```
No need to memset, calloc() already does that.

divineaugustine replied:
Thank you. These are not required anymore.

alexlarsson commented on libcomposefs/lcfs-writer.c:
```diff
@@ -471,9 +471,16 @@ struct lcfs_node_s *lcfs_node_new(void)
+	node->delayedDigestCalculation = false;
+	node->delayedInline = false;
+	node->delayedPayload = false;
+	node->delayedResetDigest = false;
```
Additionally, we generally rely on the zero initializing to do all these. Only non-zero initializers are needed.

divineaugustine replied:
This is not required anymore.

alexlarsson commented on libcomposefs/lcfs-writer.c:
```diff
@@ -590,7 +635,7 @@ static int read_content(int fd, size_t size, uint8_t *buf)
 }
 
-static void digest_to_path(const uint8_t *csum, char *buf)
+void lcfs_digest_to_path(const uint8_t *csum, char *buf)
```
This is maybe not a great function to make public it is very easy to use to shoot yourself in the foot, as it requires the caller to allocate space and whatnot. I think its better to move to lcfs-utils.h or something so that it can be used in mkcomposefs without needing to be in the public library ABI forever.

divineaugustine replied:
With the latest change, this remains as an internal function.

alexlarsson commented on libcomposefs/lcfs-writer.h:
```diff
@@ -37,6 +37,7 @@ enum {
 	LCFS_BUILD_NO_INLINE = (1 << 4),
 	LCFS_BUILD_USER_XATTRS = (1 << 5), /* Only read user.* xattrs */
 	LCFS_BUILD_BY_DIGEST = (1 << 6), /* Refer to basedir files by fs-verity digest */
+	LCFS_BUILD_DELAYED_DIGEST = (1 << 7), /* Refer to basedir files by fs-verity digest */
```
Comment is wrong

divineaugustine replied:
This is not required anymore

alexlarsson commented on libcomposefs/lcfs-writer.c:
```diff
@@ -564,7 +571,45 @@ int lcfs_node_set_fsverity_from_fd(struct lcfs_node_s *node, int fd)
+	return node->delayedInline;
+}
+
+bool lcfs_get_reset_digest(struct lcfs_node_s *node)
```
Why is this not called get_delayed_reset_digest?

divineaugustine replied:
That was a miss. These changes are not required anymore

alexlarsson commented on libcomposefs/lcfs-writer.c:
```diff
@@ -564,7 +571,45 @@ int lcfs_node_set_fsverity_from_fd(struct lcfs_node_s *node, int fd)
+	return node->delayedResetDigest;
+}
+
+int lcfs_read_content(int fd, size_t size, uint8_t *buf)
```
This too (like digest_to_path) should be an internal function, not a public library  API that we support forever.

divineaugustine replied:
This remains as an internal function as per the new changes.

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -760,46 +762,27 @@ static int mkdir_parents(const char *pathname, int mode)
-		if (bytes_read == 0)
+	do 
+	{
+		ret = copy_file_range(sfd, NULL, dfd, NULL, len, 0);
```
I like this change, but it should be a separate independent commit with commit messages describing it.

divineaugustine replied:
Added a separate commit just for this change

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+{
+	struct ParallelCopyInfo *ret = 
+		calloc(1, sizeof( struct ParallelCopyInfo));
+	ret->next 		= NULL;
```
calloc already NULLed these.

divineaugustine replied:
Changed this to an array instead of linked list. This code is not valid anymore.

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+
+static struct ParallelCopyInfo* new_parallel_copy_info(void)
+{
+	struct ParallelCopyInfo *ret = 
```
Need oom() check.

divineaugustine replied:
added checks in all the allocations.

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+{
+	if(node) 
+	{
+		cleanup_copy_node(node->next);
```
This is a non-tailcall recursion, that is going to use up a lot of stack.
This should just be a regular loop over the nodes instead.

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+		   		lcfs_node_get_payload(node) != NULL) 
+	{
+		current->node 		= node;
+		current->parentPath = strdup(path);
```
check strdup return value (and in other strdup calls)

divineaugustine replied:
Corrected

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+		const size_t n_children = lcfs_node_get_n_children(node);
+		for (size_t i = 0; i < n_children; i++)
+		{
+			current = construct_copy_data(lcfs_node_get_child(node, i), 
```
No error checking here.

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+	}	
+}
+
+static struct ParallelCopyInfo* construct_copy_data( struct lcfs_node_s* node, 
```
I don't like how this uses a current argument which is returned changed. This will just cause problems when you want to return an error or something like that. Instead create a state structure for the parallel work and put the current pointer in it, and then pass this state to all functions and modify it, then use the return value for error reporting.

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+
+	for( int i = 0; i< nrThreads; i++) 
+		pthread_join(threads[i], NULL);
+
```
I think this should create n_threads - 1 thread, and then call thread_proc on the main thread as well. Then we can efficiently handle the n_thread == 1 case without creating any threads.

divineaugustine replied:
I liked this proposal. I changed it.

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -1046,6 +1337,7 @@ int main(int argc, char **argv)
 	long min_version = 0;
 	long max_version = 0;
 	char *end;
+	int threads = 1;
```
I think the default should be n_cpus, nobody is going to use this otherwise.

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+				tm.tm_hour, tm.tm_min, tm.tm_sec, msg);
+}
+
+struct ParallelCopyInfo
```
Really, i think it makes more sense for this to be an array instead of a linked list. With the change to a state struct to keep the current pointer that should be an easy conversion to make. There shouldn't be a need to allocate a ton of individual ParallelCopyInfo objects.

divineaugustine replied:
Changed it to array

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,6 +876,307 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+			return r;
+	}
+
+	return 0;
```
In fact, I don't think you even need LCFS_BUILD_DELAYED_DIGEST. Just strip out LCFS_BUILD_BY_DIGEST 
 and LCFS_BUILD_COMPUTE_DIGEST from the call to lcfs_load_node_from_file()  and always pass NO_INLINE.
That will have the same effect as DELAYED_DIGEST, and then you just call set_from_content() with the original build flags.

divineaugustine replied:
Thank you for this suggestion. This now makes the library code simpler. I also had the same thought in my mind about duplication but i was not so sure about the flags.

alexlarsson commented:
So, lots of comments. Make sure you read them all first, because some will make the other ones perhaps not needed. In particular, the propsed change to lcfs_node_set_from_content() will make most of the library API changes unnecessary.

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -1181,14 +1481,32 @@ int main(int argc, char **argv)
-		    fill_store(root, src_path, digest_store_path) < 0)
-			err(EXIT_FAILURE, "cannot fill store");
+		fflush(stdout); 	
+		if(threads > 1)
```
Yes i totally agree with you. mkcomposefs should by default be multi-threaded in my opinion. I wanted to keep the existing flow untouched in this PR and get your feedback about threading before making threads default.

divineaugustine commented:
> So, lots of comments. Make sure you read them all first, because some will make the other ones perhaps not needed. In particular, the propsed change to lcfs_node_set_from_content() will make most of the library API changes unnecessary.

Thank you for the comments. I have addressed those and also squashed the commits to only 3 separate ones as you suggested. As i don´t know who should resolve the comments, i just added reply to those.

alexlarsson commented on libcomposefs/lcfs-writer.c:
```diff
@@ -606,6 +606,62 @@ static void digest_to_path(const uint8_t *csum, char *buf)
 }
 
+bool lcfs_node_set_from_content(struct lcfs_node_s *node, int dirfd,
+				const char *fname, int buildflags)
```
All other functions return int, where 0 == success and < 0 is error. I think we should keep doing that.

divineaugustine replied:
fixed

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -760,46 +760,25 @@ static int mkdir_parents(const char *pathname, int mode)
-		if (bytes_read == 0)
+	do {
+		ret = copy_file_range(sfd, NULL, dfd, NULL, len, 0);
+		if (ret == -1)
```
I wonder also, goes glibc have a fallback for older kernels for this?

alexlarsson replied:
Yeah, seems like it.

divineaugustine replied:
EINTR is not mentioned as a possible error for copy_file_range. Should that be handled?

alexlarsson replied:
I'm not sure if the syscall is guaranteed to never get EINTR (although I doubt it), but the glibc fallback implementation clearly doesn't handle it. So, yes, we need to handle EINTR.

divineaugustine replied:
Fixed. Thank you

giuseppe replied:
I think we still need a fallback to the older code, since `copy_file_range` doesn't seem to work across different file systems:

```
$ ./tools/mkcomposefs --digest-store=/tmp/store  tests /dev/null
mkcomposefs: cannot fill store: Invalid cross-device link
```

divineaugustine replied:
Do you suggest checking for error code ```EXDEV``` and fallback to previous implementation of read and write or totally replacing copy_file_range with read/write? copy_file_range is already a fallback for soft link attempt. Overall performance gain in my case was ~1% with this api change(i didn't run a lot of bench marks just for this).

giuseppe replied:
> copy_file_range is already a fallback for soft link attempt.

that still fails with different file systems; and I think it is good to keep the new code as it is faster.

I don't think we can add the fallback for any error as the fds could be already used, but `EXDEV` looks safe.

cgwalters replied:
It's worth using existing codebases as a reference for which errno values to use as fallback:

- Rust: https://github.com/rust-lang/rust/blob/0e5f5207881066973486e6a480fa46cfa22947e9/library/std/src/sys/pal/unix/kernel_copy.rs#L622
- libglnx: https://gitlab.gnome.org/GNOME/libglnx/-/blob/202b294e6079e23242e65e0426f8639841d1210b/glnx-fdio.c#L846 (which was long ago forked from systemd):
- systemd: https://github.com/systemd/systemd/blob/e71b40fd0026c0884ca26eb4f0a9fbe4d9285cfa/src/shared/copy.c#L338

divineaugustine replied:
Thank you for the links. Reading through https://lwn.net/Articles/846403/ i think it is safer to rollback `copy_file_range` to older `read/write` or perhaps address this as a separate PR as this does not have a direct dependency to the original multi-threading change. What do you say? I am very new to linux and i was looking for better apis to copy a file without user mode buffer.

divineaugustine commented on libcomposefs/lcfs-writer.c:
```diff
@@ -606,6 +606,62 @@ static void digest_to_path(const uint8_t *csum, char *buf)
+				const char *fname, int buildflags)
+{
+	if (node == NULL)
+		return false;
```
fixed

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+	}
+	header->items[header->count].node = node;
+	header->items[header->count].path = strdup(path);
+	if (header->items[header->count].path == NULL) {
```
Should set errno

divineaugustine replied:
fixed

alexlarsson reviewed (changes requested):
Various nitpicks, but overall this looks good to me.

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+	char *path;
+};
+
+struct work_item_header {
```
renamed to work_collection

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+		return false;
+
+	if (header->count == header->capacity) {
+		header->capacity = header->count == 0 ? 16 : header->capacity * 2;
```
fixed

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+	int capacity;
+};
+
+static bool add_to_work_collection(struct work_item_header *header,
```
fixed

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+
+		header->items = new_children;
+	}
+	header->items[header->count].node = node;
```
added ref and unref

alexlarsson replied:
These could do:
	collection->items[collection->count].node = lcfs_node_ref(node);
Instead of ref:ing on a separate line

divineaugustine replied:
fixed

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -1160,12 +1426,20 @@ int main(int argc, char **argv)
+		// Digest calculation and inline will be done in parallel
+		buildflags &= ~LCFS_BUILD_COMPUTE_DIGEST;
+		buildflags &= ~LCFS_BUILD_BY_DIGEST;
+		buildflags |= LCFS_BUILD_NO_INLINE;
```
fixed

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+	const int thread_count = requested_threads - 1;
+	if (thread_count >= 1) {
+		threads = calloc(thread_count, sizeof(pthread_t));
+		if (threads == NULL)
```
fixed

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+			return false;
+
+		for (int i = 0; i < thread_count; i++) {
+			pthread_create(&threads[i], NULL, thread_proc, &thread_info);
```
fixed

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+
+	if (thread_count >= 1) {
+		for (int i = 0; i < thread_count; i++)
+			pthread_join(threads[i], NULL);
```
fixed

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+			return 0;
+
+		if (!item->node)
+			return 0;
```
yes. it is an error if this happens. so i have added a cancel and return in case of empty node.

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -872,43 +875,283 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+
+		free(threads);
+	}
+	return !iterator.cancel_request;
```
request cancel is changed to also send error code. so it is set as errorno before returning from this function.

divineaugustine commented:
> Various nitpicks, but overall this looks good to me.

Thank you for reviewing it. I have addressed those and pushed the changes.

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,43 +877,318 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+	thread_info.collection = collection;
+	thread_info.iterator = &iterator;
+
+	pthread_t *threads = NULL;
```
I would use:
cleanup_free phread_t *threads = NULL;

then you can drop all calls to free(threads)

divineaugustine replied:
fixed

alexlarsson commented on tools/mkcomposefs.c:
```diff
@@ -893,43 +877,318 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+
+static int process_copy(struct work_item *item, void *digest_store_path)
+{
+	if (copy_file_with_dirs_if_needed(item->path, (const char *)digest_store_path,
```
I don't think you need the if here. You can just: "return copy_file_with_dirs_if_needed(...);"

divineaugustine replied:
that was a miss. fixed

alexlarsson commented:
Some more minor feedback, but I would like @cgwalters or @giuseppe to also review this.

alexlarsson commented:
Also, if you're able to talk about it, I'd love to hear what you're using composefs for. Its good to know what you're users are doing.

divineaugustine commented on tools/mkcomposefs.c:
```diff
@@ -893,43 +877,318 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
+static void request_cancel(struct work_item_iterator *iterator, int errorcode)
+{
+	pthread_mutex_lock(&(iterator->mutex_node_iterator));
+	iterator->cancel_request = true;
```
fixed

divineaugustine commented:
> Also, if you're able to talk about it, I'd love to hear what you're using composefs for. Its good to know what your users are doing.

 Can you give some details @https://github.com/r0l1 ? He is the right person to comment on this.

giuseppe reviewed (commented):
left a comment, could you please sign the commits with your real name?

divineaugustine commented:
> left a comment, could you please sign the commits with your real name?

My real name is Divin and i have signed off the commits with it. Did you mean including initials Divin OA or even initials expanded Divin Ookken Athappan?

giuseppe commented:
> My real name is Divin and i have signed off the commits with it. Did you mean including initials Divin OA or even initials expanded Divin Ookken Athappan?

yes please use the expanded version

cgwalters commented:
Yes, separate PRs for things like this that are logically separate please.

> Thank you for the links. Reading through 
> https://lwn.net/Articles/846403/ i think it is safer to rollback 
> `copy_file_range` to older `read/write` or perhaps address this as a 
> separate PR as this does not have a direct dependency to the original 
> multi-threading change. What do you say? I am very new to linux and i 
> was looking for better apis to copy a file without user mode buffer.

divineaugustine commented:
> Yes, separate PRs for things like this that are logically separate please.

I will create a separate PR for it. Should i also squash up the current commits in this PR to just 2 ?

cgwalters commented:
Yes, I think one or two commits here is best.

cgwalters commented:
I'm assuming your goal is faster `mkcomposefs` overall (on a server side build system, right?).  Using multiple threads is one way to do that.  However, another way is caching - ostree does this thing where the build system can retain a cache of (device, inode number) :arrow_right:  checksum; in theory we could offer the same.

The downside is of course that if something happens to mutate a file in the cache, you get incorrect checksums.  But this is pretty easily mitigated, and actually a great thing with composefs is such a failure should be detected at runtime.

(These are complementary approaches to be clear; threading also makes sense)

divineaugustine commented:
> Yes, I think one or two commits here is best.



> Yes, I think one or two commits here is best.

Squashed the commits to just 2 (one for library and one for mkcomposefs) and siged it off with my full name. Added an issue #274  for `copy_file_range`

alexlarsson commented:
@r0l1 That sounds cool! Nice that it works for you.

divineaugustine commented:
> I'm assuming your goal is faster `mkcomposefs` overall (on a server side build system, right?). Using multiple threads is one way to do that. However, another way is caching - ostree does this thing where the build system can retain a cache of (device, inode number) ➡️ checksum; in theory we could offer the same.
> 
> The downside is of course that if something happens to mutate a file in the cache, you get incorrect checksums. But this is pretty easily mitigated, and actually a great thing with composefs is such a failure should be detected at runtime.
> 
> (These are complementary approaches to be clear; threading also makes sense)

Thank you for these alternative-approaches. My objective was to make mkcomposefs run fast in any machine as part of our build process. It was simple and straight forward for me to add threads in mkcomposefs which does not require any facilitation from the end user. I am new to linux tooling and I would require some consultations here to fully understand these alternative approaches.

alexlarsson commented:
> I'm assuming your goal is faster `mkcomposefs` overall (on a server side build system, right?). Using multiple threads is one way to do that. However, another way is caching - ostree does this thing where the build system can retain a cache of (device, inode number) ➡️ checksum; in theory we could offer the same.

This only works though if you image compose produces hardlinked copies of files you previously committed (otherwise the device/inode of the new files will not match something old in the cache). This happens regularly in ostree because of how it is set up, but that isn't necessarily always true, and it causes some level of extra pain to enforce when you build using ostree. (With things like rofiles-fuse, etc.)

Not saying it is bad, but its not always applicable.

But anyway, if you *do* use a setup like that where you end up with hardlinked files, then the best way to store and quickly access the cached digest is probably to just enable fs-verity on these files. In other words, we should probably have lcfs_node_set_from_content() start by asking the kernel for the fs-verity digest in case it is set, rather than computing it ourselves.

cgwalters commented:
> In other words, we should probably have lcfs_node_set_from_content() start by asking the kernel for the fs-verity digest in case it is set, rather than computing it ourselves.

Yes agreed, this is an obvious step.

alexlarsson commented:
I think this looks good enough now. I looked through it again and did some local tests, seems to work well, and the public library APIs are sane.
